### PR TITLE
Abstract away model-filemodel and add mutable stores

### PR DESCRIPTION
**Pull request by peppy**
- Lays groundwork for implementing arbitrary database representations of archives (Beatmaps, Skins etc.).
- Centralises import logic so it is no longer scattered between three or four classes.
- Fixes a few performance issues due to incorrect usage of `IQueryable<>`.
- Moves archive handling (`ArchiveReader`) out of `Beatmap` namespace, as there is no reason for it to remain there when it can be used elsewhere.

### CODE REVIEW

File: osu.Game/Database/ArchiveModelManager.cs [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+        public void Delete(TModel item)
+        {
+            using (var usage = ContextFactory.GetForWrite())
+            {
+                var context = usage.Context;
+
+                context.ChangeTracker.AutoDetectChangesEnabled = false;
+
+                // re-fetch the model on the import context.
+                var foundModel = queryModel().Include(s => s.Files).ThenInclude(f => f.FileInfo).First(s => s.ID == item.ID);
```
Comment (smoogipoo) [inline osu.Game/Database/ArchiveModelManager.cs]: Is this going to be safe? It looks like `queryModel()` always returns read-only contexts, which do not track whether they've written.

Comment (smoogipoo) [reply osu.Game/Database/ArchiveModelManager.cs]: I guess it is since ModelStore.Delete is run on the write context?

Comment (peppy) [reply osu.Game/Database/ArchiveModelManager.cs]: We're inside a write context. Remember the write context is just using the current thread's read context, so this is always the same context.

Comment (smoogipoo) [reply osu.Game/Database/ArchiveModelManager.cs]: The second part of that was the more critical point - non-write context don't track whether they've written anything.

Comment (peppy) [reply osu.Game/Database/ArchiveModelManager.cs]: I'm not sure what you mean by that.

Comment (peppy) [reply osu.Game/Database/ArchiveModelManager.cs]: See line 176 – we're already within a write usage.

Comment (smoogipoo) [reply osu.Game/Database/ArchiveModelManager.cs]: Right, I see.

File: osu.Game/Database/ArchiveModelManager.cs [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+
+            return fileInfos;
+        }
+
+        /// <summary>
+        /// Create a barebones model from the provided archive.
+        /// Actual expensive population should be done in <see cref="Populate"/>; this should just prepare for duplicate checking.
+        /// </summary>
+        /// <param name="archive"></param>
+        /// <returns></returns>
```
Comment (smoogipoo) [inline osu.Game/Database/ArchiveModelManager.cs]: Unpopulated tags.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**